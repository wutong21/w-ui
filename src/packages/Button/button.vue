<template>
  <button
    :disabled="buttonDisabled || loading"
    :autofocus="autofocus"
    :type="nativeType"
    :class="[
      type ? 'el-button--' + type : '',
      size ? 'el-button--' + size : '',
      {
        'is-disabled': buttonDisabled,
        'is-loading': loading,
        'is-plain': plain,
        'is-radius': radius,
        'is-icon': icon || iconBorder,
        'is-icon-border': iconBorder,
        'is-icon-text': $slots.default && icon
      }
    ]"
    :style="{
      minWidth: minWidth && `${minWidth}px`,
      maxWidth: maxWidth && `${maxWidth}px`
    }"
    :title="labelTitle"
    class="el-button"
    @click="handleClick"
    @mouseover.self="handleMouseover"
  >
    <!-- <i class="el-icon-loading" v-if="loading" @click="handleInnerClick"></i> -->
    <el-load-icon
      v-if="loading"
      class="h-icon-loading"
      @click="handleInnerClick"
    />
    <i v-if="icon && !loading" :class="icon" />
    <i
      v-if="iconBorder && !loading"
      :class="iconBorder"
      @click="handleInnerClick"
    />
    <!-- <i :class="outicon" v-if="outicon" @click="handleInnerClick"></i> -->
    <span v-if="$slots.default" @click="handleInnerClick">
      <slot />
    </span>
  </button>
</template>
<script>
export default {
  name: 'ElButton',
  props: {
    minWidth: {
      type: Number,
      default: null
    },
    maxWidth: {
      type: Number,
      default: null
    },
    type: {
      type: String,
      default: 'default'
    },
    size: {
      type: String,
      default: null
    },
    icon: { type: String, default: null },
    nativeType: {
      type: String,
      default: 'button'
    },
    loading: {
      type: Boolean,
      default: null
    },
    disabled: {
      type: Boolean,
      default: null
    },
    iconBorder: {
      type: String,
      default: null
    },
    plain: {
      type: Boolean,
      default: null
    },
    autofocus: {
      type: Boolean,
      default: null
    },
    radius: {
      type: Boolean,
      default: null
    }
  },

  data() {
    return {
      labelTitle: ''
    };
  },

  computed: {
    buttonDisabled() {
      return this.disabled || (this.elForm || {}).disabled;
    }
  },

  methods: {
    handleInnerClick(e) {
      if (this.disabled) {
        e.stopPropagation();
      }
    },

    handleClick(evt) {
      this.$emit('click', evt);
    },

    handleMouseover(e) {
      const t = e.target;
      // 如果设置了默认的title就用默认的
      if (this.$attrs.title) return;
      this.labelTitle = t.scrollWidth > t.offsetWidth ? t.innerText : '';
    }
  }
};
</script>
