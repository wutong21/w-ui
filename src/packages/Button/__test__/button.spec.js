import { mount } from "@vue/test-utils";
import { describe, test, expect } from "vitest";
import Button from "../button.vue";

describe("Button", () => {
  test("snapshot test", () => {
    const wrapper = mount(Button);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
