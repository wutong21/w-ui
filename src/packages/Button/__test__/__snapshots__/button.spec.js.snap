// Vitest Snapshot v1

exports[`Button > snapshot test 1`] = `
"<button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--default\\">
  <!---->
  <!---->
  <!---->
  <!----></button>"
`;

exports[`Button snapshot test 1`] = `"<div><button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--primary\\"><!----> <!----> <!----> <span>主要按钮</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--default\\"><!----> <!----> <!----> <span>白色次按钮</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--primary is-radius\\"><!----> <!----> <!----> <span>圆角按钮</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--text\\"><!----> <!----> <!----> <span>文字按钮</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--link\\"><!----> <!----> <!----> <span>文字链接</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--default is-icon\\"><!----> <i class=\\"h-icon-search\\"></i> <!----> <!----></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--default is-icon is-icon-text\\"><!----> <i class=\\"h-icon-search\\"></i> <!----> <span>Search</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--primary is-icon is-icon-text\\"><!----> <i class=\\"h-icon-search\\"></i> <!----> <span>Search</span></button> <button disabled=\\"disabled\\" type=\\"button\\" title=\\"\\" class=\\"el-button el-button--primary is-loading\\"><span class=\\"h-an-icon-loading h-icon-loading\\"><span class=\\"path1\\"></span> <span class=\\"path2\\"></span> <span class=\\"path3\\"></span> <span class=\\"path4\\"></span></span> <!----> <!----> <span>加载中</span></button> <button disabled=\\"disabled\\" type=\\"button\\" title=\\"\\" class=\\"el-button el-button--text is-loading\\"><span class=\\"h-an-icon-loading h-icon-loading\\"><span class=\\"path1\\"></span> <span class=\\"path2\\"></span> <span class=\\"path3\\"></span> <span class=\\"path4\\"></span></span> <!----> <!----> <span>文字按钮</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--primary el-button--large\\"><!----> <!----> <!----> <span>large</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--primary\\"><!----> <!----> <!----> <span>default</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--primary el-button--small\\"><!----> <!----> <!----> <span>small</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--primary el-button--mini\\"><!----> <!----> <!----> <span>mini</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--success\\"><!----> <!----> <!----> <span>成功按钮</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--warning\\"><!----> <!----> <!----> <span>警告按钮</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--danger\\"><!----> <!----> <!----> <span>危险按钮</span></button> <button type=\\"button\\" title=\\"\\" class=\\"el-button el-button--info\\"><!----> <!----> <!----> <span>信息按钮</span></button></div>"`;
